File: ai-service/sentiment_analyzer.py
"""
============================================
PerfumeShop V18 - Sentiment Analyzer
Review sentiment analysis using keyword-based approach
(with jieba for Chinese text segmentation)
============================================
"""
import logging
import re
from typing import Dict, List, Any

logger = logging.getLogger(__name__)

try:
    import jieba
    JIEBA_AVAILABLE = True
except ImportError:
    JIEBA_AVAILABLE = False
    logger.warning("jieba not installed, using basic word matching")


class SentimentAnalyzer:
    """Analyzes sentiment of product review texts."""
    
    # Sentiment word dictionaries (Chinese)
    POSITIVE_WORDS = {
        '喜欢', '不错', '好闻', '香', '推荐', '满意', '棒', '赞', '完美', '惊喜',
        '优雅', '清新', '持久', '高级', '自然', '舒服', '柔和', '温柔', '迷人',
        '好评', '超值', '回购', '性价比', '经典', '独特', '精致', '好', '爱',
        'nice', 'love', 'good', 'great', 'beautiful', 'excellent', 'perfect',
        'wonderful', 'amazing', 'fantastic', 'best', 'lovely'
    }
    
    NEGATIVE_WORDS = {
        '失望', '不好', '难闻', '刺鼻', '不持久', '太淡', '太浓', '不值', '差',
        '后悔', '过敏', '头晕', '恶心', '酒精', '假货', '劣质', '粗糙', '呛',
        'bad', 'poor', 'terrible', 'awful', 'worst', 'horrible', 'dislike',
        'cheap', 'fake', 'weak', 'strong', 'headache', 'allergy'
    }
    
    INTENSIFIERS = {
        '非常': 2.0, '特别': 2.0, '超级': 2.5, '极其': 2.5, '太': 1.8,
        '很': 1.5, '挺': 1.3, '有点': 0.7, '稍微': 0.6, '略微': 0.6,
        'very': 2.0, 'extremely': 2.5, 'really': 1.8, 'quite': 1.5,
        'so': 1.5, 'absolutely': 2.5
    }
    
    NEGATORS = {'不', '没', '无', '非', 'not', 'no', "don't", "doesn't", "didn't"}
    
    def __init__(self):
        self._ready = True
        
    def is_ready(self) -> bool:
        return self._ready
    
    def analyze(self, text: str) -> Dict[str, Any]:
        """
        Analyze sentiment of a single text.
        
        Returns:
            Dict with sentiment ('positive'/'negative'/'neutral'),
            score (-1.0 to 1.0), keywords, and confidence.
        """
        text = text.strip()
        if not text:
            return {
                'sentiment': 'neutral',
                'score': 0.0,
                'confidence': 1.0,
                'keywords': [],
                'summary': ''
            }
        
        # Tokenize
        words = self._tokenize(text)
        
        # Count sentiments
        pos_count = 0
        neg_count = 0
        pos_keywords = []
        neg_keywords = []
        
        for i, word in enumerate(words):
            # Check for negator before sentiment word
            negator_before = False
            if i > 0 and words[i-1] in self.NEGATORS:
                negator_before = True
            
            # Check for intensifier
            intensifier = 1.0
            if i > 0 and words[i-1] in self.INTENSIFIERS:
                intensifier = self.INTENSIFIERS[words[i-1]]
            
            if word in self.POSITIVE_WORDS:
                if negator_before:
                    neg_count += intensifier
                    neg_keywords.append(f"不{word}")
                else:
                    pos_count += intensifier
                    pos_keywords.append(word)
            elif word in self.NEGATIVE_WORDS:
                if negator_before:
                    pos_count += intensifier * 0.7
                    pos_keywords.append(f"不{word}")
                else:
                    neg_count += intensifier
                    neg_keywords.append(word)
        
        # Calculate score
        total = pos_count + neg_count
        if total == 0:
            score = 0.0
            sentiment = 'neutral'
        else:
            score = (pos_count - neg_count) / (pos_count + neg_count)
        
        # Determine sentiment
        if score > 0.2:
            sentiment = 'positive'
        elif score < -0.2:
            sentiment = 'negative'
        else:
            sentiment = 'neutral'
        
        # Confidence based on keyword count
        keyword_count = len(pos_keywords) + len(neg_keywords)
        confidence = min(1.0, keyword_count / 5.0) if keyword_count > 0 else 0.3
        
        # Generate summary
        summary = self._generate_summary(sentiment, pos_keywords[:5], neg_keywords[:5])
        
        return {
            'sentiment': sentiment,
            'score': round(score, 3),
            'confidence': round(confidence, 3),
            'keywords': {
                'positive': pos_keywords[:5],
                'negative': neg_keywords[:5]
            },
            'summary': summary
        }
    
    def _tokenize(self, text: str) -> List[str]:
        """Tokenize text into words/phrases."""
        # Remove special characters but keep Chinese
        text = re.sub(r'[^\u4e00-\u9fff\w\s]', ' ', text)
        
        if JIEBA_AVAILABLE and re.search(r'[\u4e00-\u9fff]', text):
            # Use jieba for Chinese text
            return [w.strip() for w in jieba.cut(text) if w.strip()]
        else:
            # Simple whitespace split for English
            return text.lower().split()
    
    def _generate_summary(self, sentiment: str, pos_words: List[str], neg_words: List[str]) -> str:
        """Generate a human-readable summary."""
        if sentiment == 'positive':
            if pos_words:
                return f"用户评价积极，关键词: {', '.join(pos_words[:3])}"
            return "用户评价偏正面"
        elif sentiment == 'negative':
            if neg_words:
                return f"用户评价消极，关键词: {', '.join(neg_words[:3])}"
            return "用户评价偏负面"
        else:
            return "用户评价中性"

File: src/PerfumeShop.Core/AI/SentimentAnalyzer.cs
using System.Text.RegularExpressions;
using JiebaNet.Segmenter;

namespace PerfumeShop.Core.AI;

/// <summary>
/// 基于词典的情感分析器（移植自 ai-service/sentiment_analyzer.py）
/// 使用 jieba.NET 进行中文分词，内置情感词库 + 否定词处理 + 程度副词
/// </summary>
public class SentimentAnalyzer : ISentimentAnalyzer
{
    private static readonly HashSet<string> PositiveWords = new(StringComparer.OrdinalIgnoreCase)
    {
        "喜欢", "不错", "好闻", "香", "推荐", "满意", "棒", "赞", "完美", "惊喜",
        "优雅", "清新", "持久", "高级", "自然", "舒服", "柔和", "温柔", "迷人",
        "好评", "超值", "回购", "性价比", "经典", "独特", "精致", "好", "爱",
        "nice", "love", "good", "great", "beautiful", "excellent", "perfect",
        "wonderful", "amazing", "fantastic", "best", "lovely"
    };

    private static readonly HashSet<string> NegativeWords = new(StringComparer.OrdinalIgnoreCase)
    {
        "失望", "不好", "难闻", "刺鼻", "不持久", "太淡", "太浓", "不值", "差",
        "后悔", "过敏", "头晕", "恶心", "酒精", "假货", "劣质", "粗糙", "呛",
        "bad", "poor", "terrible", "awful", "worst", "horrible", "dislike",
        "cheap", "fake", "weak", "strong", "headache", "allergy"
    };

    private static readonly Dictionary<string, double> Intensifiers = new(StringComparer.OrdinalIgnoreCase)
    {
        ["非常"] = 2.0, ["特别"] = 2.0, ["超级"] = 2.5, ["极其"] = 2.5, ["太"] = 1.8,
        ["很"] = 1.5, ["挺"] = 1.3, ["有点"] = 0.7, ["稍微"] = 0.6, ["略微"] = 0.6,
        ["very"] = 2.0, ["extremely"] = 2.5, ["really"] = 1.8, ["quite"] = 1.5,
        ["so"] = 1.5, ["absolutely"] = 2.5
    };

    private static readonly HashSet<string> Negators = new(StringComparer.OrdinalIgnoreCase)
    {
        "不", "没", "无", "非", "not", "no", "don't", "doesn't", "didn't"
    };

    private static readonly Regex ChinesePattern = new(@"[\u4e00-\u9fff]", RegexOptions.Compiled);
    private static readonly Regex CleanPattern = new(@"[^\u4e00-\u9fff\w\s]", RegexOptions.Compiled);

    private readonly JiebaSegmenter? _segmenter;

    public bool IsReady { get; }

    public SentimentAnalyzer()
    {
        try
        {
            _segmenter = new JiebaSegmenter();
            IsReady = true;
        }
        catch
        {
            IsReady = true; // 降级到空格分词也能工作
        }
    }

    public SentimentResult Analyze(string text)
    {
        text = (text ?? "").Trim();
        if (string.IsNullOrEmpty(text))
        {
            return new SentimentResult
            {
                Label = "neutral", Score = 0.0, Confidence = 1.0,
                Keywords = new SentimentKeywords(), Summary = ""
            };
        }

        var words = Tokenize(text);

        double posCount = 0, negCount = 0;
        var posKeywords = new List<string>();
        var negKeywords = new List<string>();

        for (int i = 0; i < words.Count; i++)
        {
            var word = words[i];
            bool negatorBefore = i > 0 && Negators.Contains(words[i - 1]);
            double intensifier = 1.0;
            if (i > 0 && Intensifiers.TryGetValue(words[i - 1], out var intens))
                intensifier = intens;

            if (PositiveWords.Contains(word))
            {
                if (negatorBefore)
                {
                    negCount += intensifier;
                    negKeywords.Add($"不{word}");
                }
                else
                {
                    posCount += intensifier;
                    posKeywords.Add(word);
                }
            }
            else if (NegativeWords.Contains(word))
            {
                if (negatorBefore)
                {
                    posCount += intensifier * 0.7;
                    posKeywords.Add($"不{word}");
                }
                else
                {
                    negCount += intensifier;
                    negKeywords.Add(word);
                }
            }
        }

        double total = posCount + negCount;
        double score = total == 0 ? 0.0 : (posCount - negCount) / total;

        string label = score > 0.2 ? "positive" : score < -0.2 ? "negative" : "neutral";

        int keywordCount = posKeywords.Count + negKeywords.Count;
        double confidence = keywordCount > 0 ? Math.Min(1.0, keywordCount / 5.0) : 0.3;

        return new SentimentResult
        {
            Score = Math.Round(score, 3),
            Label = label,
            Confidence = Math.Round(confidence, 3),
            Keywords = new SentimentKeywords
            {
                Positive = posKeywords.Take(5).ToList(),
                Negative = negKeywords.Take(5).ToList()
            },
            Summary = GenerateSummary(label, posKeywords.Take(5).ToList(), negKeywords.Take(5).ToList())
        };
    }

    public List<SentimentResult> BatchAnalyze(IEnumerable<string> texts)
    {
        return texts.Select(Analyze).ToList();
    }

    private List<string> Tokenize(string text)
    {
        text = CleanPattern.Replace(text, " ");

        if (_segmenter != null && ChinesePattern.IsMatch(text))
        {
            return _segmenter.Cut(text)
                .Where(w => !string.IsNullOrWhiteSpace(w))
                .Select(w => w.Trim())
                .ToList();
        }

        return text.ToLower().Split(' ', StringSplitOptions.RemoveEmptyEntries).ToList();
    }

    private static string GenerateSummary(string sentiment, List<string> posWords, List<string> negWords)
    {
        if (sentiment == "positive")
            return posWords.Count > 0
                ? $"用户评价积极，关键词: {string.Join(", ", posWords.Take(3))}"
                : "用户评价偏正面";
        if (sentiment == "negative")
            return negWords.Count > 0
                ? $"用户评价消极，关键词: {string.Join(", ", negWords.Take(3))}"
                : "用户评价偏负面";
        return "用户评价中性";
    }
}
